src/blockstore/hashmap_db.zig

```diff
@@ -0,0 +1,254 @@
+            .allocator = self.allocator,
+            .data = ret,
+        };
+    }
+
+    pub fn delete(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+    ) !void {
+        const key_bytes = try cf.key().serializeAlloc(self.allocator, key);
+        defer self.allocator.free(key_bytes);
+        self.transaction_lock.lockShared();
+        defer self.transaction_lock.unlockShared();
+        _ = self.maps[cf_index].delete(self.allocator, key_bytes);
+    }
+
+    pub fn writeBatch(self: *Self) MapBatch {
+        return .{
+            .allocator = self.allocator,
+            .instructions = .{},
+        };
+    }
+
+    pub fn initBatch(self: *Self) MapBatch {
+        return MapBatch.init(self.allocator);
+    }
+
+    /// Atomicity may be violated if there is insufficient
+    /// memory to complete a PUT.
+    pub fn commit(self: *Self, batch: MapBatch) !void {
+        self.transaction_lock.lock();
+        defer self.transaction_lock.unlock();
+
+        for (batch.instructions.items) |ix| {
+            switch (ix) {
+                .put => |put_ix| {
+                    const cf_index, const key, const value = put_ix;
+                    try self.maps[cf_index].put(key, value);
+                },
+                .delete => |delete_ix| {
+                    const cf_index, const key = delete_ix;
+                    self.maps[cf_index].delete(batch.allocator, key);
+                },
+            }
+        }
+    }
+};
+
+pub const MapBatch = struct {
+    allocator: Allocator,
+    instructions: std.ArrayListUnmanaged(Instruction),
+
+    const Instruction = union(enum) {
+        put: struct { usize, []const u8, []const u8 },
+        delete: struct { usize, []const u8 },
+    };
+
+    const Self = @This();
+
+    pub fn init(allocator: Allocator) Self {
+        return .{
+            .allocator = allocator,
+            .instructions = std.ArrayListUnmanaged(Instruction){},
+        };
+    }
+
+    pub fn deinit(self: Self) void {
+        self.instructions.deinit(self.allocator);
+    }
+
+    pub fn put(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+        value: cf.Value,
+    ) !void {
+        return try self.instructions.append(self.allocator, .{ .put = .{
+            cf_index,
+            try cf.key().serializeAlloc(self.allocator, key),
+            try cf.value().serializeAlloc(self.allocator, value),
+        } });
+    }
+
+    pub fn delete(
+        self: *Self,
+        comptime cf: ColumnFamily,
+        comptime cf_index: usize,
+        key: cf.Key,
+    ) !void {
+        return try self.instructions.append(self.allocator, .{ .delete = .{
+            cf_index,
+            try cf.key().serializeAlloc(self.allocator, key),
+        } });
+    }
+};
+
+const SharedHashMap = struct {
+    allocator: Allocator,
+    map: std.StringHashMap([]const u8),
```

Comment:
Might as well make the map unmanaged if we're carrying its allocator anyways.

Reply:
cf54686a55196c62abae1a50badfb538e79a052e